sql: fix panic for UPDATES on tables with an inverted index

m-schneider opened the pull request:
Before this patch we would panic on updates to columns which had an
inverted index on them and the new column value would generate more
entries than the one being replaced.

After this patch we handle updates to indexes which generate a different
number of entries for the old and new values correctly.

Closes: #23290

Release note: None

Commit:
sql: fix panic for UPDATES on tables with an inverted index

Before this patch we would panic on updates to columns which had an
inverted index on them and the new column value would generate more
entries than the one being replaced.

After this patch we handle updates to indexes which generate a different
number of entries for the old and new values correctly.

Closes: #23290

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23318)
<!-- Reviewable:end -->

justinj commented:
Are you still planning to add some tests involving inserting random JSON data?

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/rowwriter.go, line 835 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6qfOXaXqiv51niaUSF:-L6qfOXaXqiv51niaUSG:byluqi3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbbfa2654add0140c8f5a9ad8063ee6c8de4ea/pkg/sql/sqlbase/rowwriter.go#L835)):*
> ```Go
> 	// an existing KV entry.
> 	var lastSeenOldEntry int
> 	for i, newSecondaryIndexEntry := range newSecondaryIndexEntries {
> ```

Maybe I'm misunderstanding, this loop feels a bit strange to me now - couldn't we have a case where delete an old entry (but re-add it) despite it still being present in the object? Say if we replaced `{"b": 1}` with `{"a": 2, "b": 1}`? Because the `"a": 1` entry will be compared to the existing `"b": 1` entry and delete it?

---

*[pkg/sql/sqlbase/rowwriter.go, line 845 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6qe2f69FIYNheE0EDW:-L6qe2f69FIYNheE0EDX:bepi3bl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbbfa2654add0140c8f5a9ad8063ee6c8de4ea/pkg/sql/sqlbase/rowwriter.go#L845)):*
> ```Go
> 
> 		var expValue interface{}
> 		if !bytes.Equal(newSecondaryIndexEntry.Key, oldSecondaryIndexEntry.Key) && len(oldSecondaryIndexEntry.Key) > 0 {
> ```

I think it would be worth extracting `len(oldSecondaryIndexEntry.Key) > 0` to a local variable like `hadOldEntry` or something, since it's used here and below

---

*[pkg/sql/sqlbase/rowwriter.go, line 858 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6qf1cg26r4FrGO0guS:-L6qf1cg26r4FrGO0guT:b1ciee5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbbfa2654add0140c8f5a9ad8063ee6c8de4ea/pkg/sql/sqlbase/rowwriter.go#L858)):*
> ```Go
> 
> 		// Do not update Indexes in the DELETE_ONLY state.
> 		if _, ok := ru.deleteOnlyIndex[i]; !ok {
> ```

Is this code path tested with inverted indexes anywhere?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-:-L6qg3HL1zSOMJpZpKSg:bufi8t9)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Yes, I was planning it as a stand-alone PR though.

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/rowwriter.go, line 835 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6qfOXaXqiv51niaUSF:-L6qiPP9r-V1gE7daoe2:bo3b9r6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbbfa2654add0140c8f5a9ad8063ee6c8de4ea/pkg/sql/sqlbase/rowwriter.go#L835)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Maybe I'm misunderstanding, this loop feels a bit strange to me now - couldn't we have a case where delete an old entry (but re-add it) despite it still being present in the object? Say if we replaced `{"b": 1}` with `{"a": 2, "b": 1}`? Because the `"a": 1` entry will be compared to the existing `"b": 1` entry and delete it?
</blockquote></details>

Yes that would happen in this case. To get around it we would have to do a merge of the old and new keys.

---

*[pkg/sql/sqlbase/rowwriter.go, line 845 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6qe2f69FIYNheE0EDW:-L6qm3mZf6fe56mpsOnl:bb0so8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbbfa2654add0140c8f5a9ad8063ee6c8de4ea/pkg/sql/sqlbase/rowwriter.go#L845)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think it would be worth extracting `len(oldSecondaryIndexEntry.Key) > 0` to a local variable like `hadOldEntry` or something, since it's used here and below
</blockquote></details>

That makes sense. I was thinking over the weekend, do you think it makes sense for a separate path for inverted indexes?

---

*[pkg/sql/sqlbase/rowwriter.go, line 858 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6qf1cg26r4FrGO0guS:-L6qlTwCDwx07X0t2ps0:b-9q9zw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbbfa2654add0140c8f5a9ad8063ee6c8de4ea/pkg/sql/sqlbase/rowwriter.go#L858)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Is this code path tested with inverted indexes anywhere?
</blockquote></details>

Yes, I added logic tests to specifically go through the new code paths. There's a test where a new value is shorter than an old value and vice versa.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-:-L6qlyN1kFXlI_0t6VMJ:b3k79hk)*
<!-- Sent from Reviewable.io -->

justinj commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/rowwriter.go, line 835 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6qfOXaXqiv51niaUSF:-L6sAQUcXguXYpcGq_wl:b-6ngw2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbbfa2654add0140c8f5a9ad8063ee6c8de4ea/pkg/sql/sqlbase/rowwriter.go#L835)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Yes that would happen in this case. To get around it we would have to do a merge of the old and new keys.
</blockquote></details>

Is that prohibitive? Can you add a TODO or something?

---

*[pkg/sql/sqlbase/rowwriter.go, line 845 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6qe2f69FIYNheE0EDW:-L6sAn912S0QD-gNwGwA:bdctads) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbbfa2654add0140c8f5a9ad8063ee6c8de4ea/pkg/sql/sqlbase/rowwriter.go#L845)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

That makes sense. I was thinking over the weekend, do you think it makes sense for a separate path for inverted indexes?
</blockquote></details>

Yes, I think that would be good!

---

*[pkg/sql/sqlbase/rowwriter.go, line 858 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6qf1cg26r4FrGO0guS:-L6sAsSyntnTFCV-7KHb:bdsocb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbbfa2654add0140c8f5a9ad8063ee6c8de4ea/pkg/sql/sqlbase/rowwriter.go#L858)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Yes, I added logic tests to specifically go through the new code paths. There's a test where a new value is shorter than an old value and vice versa.
</blockquote></details>

Oh I missed the `!ok`, I thought this was stuff *for* delete-only indexes :+1:

---

*[pkg/sql/sqlbase/rowwriter.go, line 878 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6sB2PvhuhcjU8vT0PE:-L6sB2PvhuhcjU8vT0PF:b-eod7iq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e979e943bf36c6cebc3a9161e5e3a4598652388b/pkg/sql/sqlbase/rowwriter.go#L878)):*
> ```Go
> 	// the old ones from the KV.
> 	for i := lastSeenOldEntry + 1; i < len(oldSecondaryIndexEntries); i++ {
> 		batch.Del(oldSecondaryIndexEntries[i].Key)
> ```

Could this result in deleting an entry we also inserted in this batch?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/rowwriter.go, line 835 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23318#-L6qfOXaXqiv51niaUSF:-L6sFwow4ReFTYFv1U63:b-a27779) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3dbbfa2654add0140c8f5a9ad8063ee6c8de4ea/pkg/sql/sqlbase/rowwriter.go#L835)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Is that prohibitive? Can you add a TODO or something?
</blockquote></details>

I think the worst scenario is if we have an object replaced by a very similar object sharing a lot of keys but shifted a bit, we'll do some more writes to the KV, so adding a merge would be an optimization for a fairly specific case.

---

*[pkg/sql/sqlbase/rowwriter.go, line 878 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23318#-L6sB2PvhuhcjU8vT0PE:-L6sJzRW0NKHdE21YKd0:byha746) ([raw file](https://github.com/cockroachdb/cockroach/blob/e979e943bf36c6cebc3a9161e5e3a4598652388b/pkg/sql/sqlbase/rowwriter.go#L878)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Could this result in deleting an entry we also inserted in this batch?
</blockquote></details>

No we won't delete anything we inserted. There are 3 paths through the code. One is where the there is the same number of old and new secondary index entries in which case it behaves like before. Then we have the case where there are less new entries than old entries which is taken care of by the check on line 840. This loop is for the case where there are more new entries which weren't processed because the loop above is bounded by the number of old entries. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23318)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -828,29 +828,54 @@ func (ru *RowUpdater) UpdateRow(
+	// an existing KV entry.
+	var lastSeenOldEntry int
 	for i, newSecondaryIndexEntry := range newSecondaryIndexEntries {
-		secondaryIndexEntry := secondaryIndexEntries[i]
```
Is there value in also renaming this to oldSecondaryIndexEntries?

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -828,29 +828,54 @@ func (ru *RowUpdater) UpdateRow(
+		var oldSecondaryIndexEntry IndexEntry
+		if len(secondaryIndexEntries) > i {
+			oldSecondaryIndexEntry = secondaryIndexEntries[i]
+			lastSeenOldEntry = i
```
will lastSeenOldEntry == len(secondaryIndexEntries) at the end of the loop?

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -828,29 +828,54 @@ func (ru *RowUpdater) UpdateRow(
+	for i := lastSeenOldEntry + 1; i < len(secondaryIndexEntries); i++ {
+		batch.Del(secondaryIndexEntries[i].Key)
+	}
+
```
Looks like you want to go through the list of indexes and depending on which one is larger take some specific actions here.

This code is getting to be quite complex which makes me wonder if it is possible to treat inverted indexes separately. Perhaps we should consider returning inverted indexes separately in encodeIndexes() above

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -828,29 +828,54 @@ func (ru *RowUpdater) UpdateRow(
+			if len(oldSecondaryIndexEntry.Key) > 0 {
+				batch.CPut(newSecondaryIndexEntry.Key, &newSecondaryIndexEntry.Value, expValue)
+			} else {
+				batch.InitPut(newSecondaryIndexEntry.Key, &newSecondaryIndexEntry.Value, true)
```
Is there a test that will break if this were a CPut?

m-schneider commented:
I think the two commits I had in the PR were getting a bit confusing. It was renamed in the combination of the two. I remerged them.

---

Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sqlbase/rowwriter.go, line 841 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r2-841:-L6wdzin-8ROz-dmSmza:bdcv6kv) ([raw file](https://github.com/cockroachdb/cockroach/blob/600053cc18dcb33e4529b556b7563ca00f1ef634/pkg/sql/sqlbase/rowwriter.go#L841)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

will lastSeenOldEntry == len(secondaryIndexEntries) at the end of the loop?
</blockquote></details>

Good point!

---

*[pkg/sql/sqlbase/rowwriter.go, line 868 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r2-868:-L6wnncO0lwefBrCawAg:b-x766ew) ([raw file](https://github.com/cockroachdb/cockroach/blob/600053cc18dcb33e4529b556b7563ca00f1ef634/pkg/sql/sqlbase/rowwriter.go#L868)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Is there a test that will break if this were a CPut?
</blockquote></details>

I added a test that would break with a CPut. I checked it before pushing. It goes back to a CPut not allowing you to write the same KV twice even if they entries are identical.

---

*[pkg/sql/sqlbase/rowwriter.go, line 878 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r2-878:-L6we_yG4DEDckBEFcuo:b-5kldxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/600053cc18dcb33e4529b556b7563ca00f1ef634/pkg/sql/sqlbase/rowwriter.go#L878)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Looks like you want to go through the list of indexes and depending on which one is larger take some specific actions here.

This code is getting to be quite complex which makes me wonder if it is possible to treat inverted indexes separately. Perhaps we should consider returning inverted indexes separately in encodeIndexes() above
</blockquote></details>

Yes, I agree it's starting to get clunky. We can treat them separately because we know that the slice index of an entry should correspond to an index and the rest of the entries will be at the end. What do you think about changing index entries to be a slice of slices so each slice index still corresponds to an index, but they could have multiple entries? Are the extra allocations worth it? Changing it that way could also help catch and prevent more panics.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23318#-:-L6wnaVZEQqrGoeye2Ye:b-xaxxev)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -828,29 +828,52 @@ func (ru *RowUpdater) UpdateRow(
-		secondaryIndexEntry := secondaryIndexEntries[i]
+		// We're making sure that if there are more new index entries for an inverted index than old ones, we will
+		// handle the checks for insertion and deletion correctly.
+		var oldSecondaryIndexEntry IndexEntry
```
Are you able to change this logic to

```
if len(oldSecondaryIndexEntries) <= i {
batch.InitPut(newSecondaryIndexEntry.Key, &newSecondaryIndexEntry.Value, true)
}
```

and then not need to make the changes below with the exception of the last bit that deletes extra old entries?

m-schneider commented:
Review status: 0 of 2 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 520 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L7R1DL6-ZjOIzrpAkLR:-L7R1DL6-ZjOIzrpAkLS:b-h95rsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d8c3b100b20feee24dc683f55dc4bcd79e63d4f/pkg/sql/logictest/testdata/logic_test/inverted_index#L520)):*
> ```
> 
> query IT
> SELECT * from update_test where j @> '["b"]';
> ```

This test was failing with the previous code.

---

*[pkg/sql/sqlbase/rowwriter.go, line 837 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r4-837:-L75jobQC_deeIwfBkZb:b-k4ez7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fd1ea4eebb73f35c52bc72b844fec6ed8f106bf/pkg/sql/sqlbase/rowwriter.go#L837)):*
> if len(oldSecondaryIndexEntries) <= i {
> batch.InitPut(newSecondaryIndexEntry.Key, &newSecondaryIndexEntry.Value, true)
> }
I don't think so because there could be less new entries than old

---

*[pkg/sql/sqlbase/rowwriter.go, line 854 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L7R1R40VqLsnE0q4AAV:-L7R1R40VqLsnE0q4AAW:bw4euk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d8c3b100b20feee24dc683f55dc4bcd79e63d4f/pkg/sql/sqlbase/rowwriter.go#L854)):*
> ```Go
> 
> 		// Do not update Indexes in the DELETE_ONLY state.
> 		if _, ok := ru.deleteOnlyIndex[i]; !ok {
> ```

I'm not quite sure what to do about this, because if we have multiple inverted indexes all of their entries are going to be mixed at the end.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
@m-schneider the current model is pretty badly broken. I think you're going to have to move the inverted indexes out and return them separately as a list of lists.

vivekmenezes commented:
Another option is to treat this like the insert case where we filter out
the delete only indexes and run through this and then later process the
delete only indexes separately. I still prefer us changing the signature of
the encoding indexes function to return the inverted indexes separately but
since we're so close to the end here we can do the above

On Mon, Mar 12, 2018, 5:08 PM Masha Schneider <notifications@github.com>
wrote:

> Review status: 0 of 2 files reviewed at latest revision, 10 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *pkg/sql/logictest/testdata/logic_test/inverted_index, line 520 at r5
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L7R1DL6-ZjOIzrpAkLR:-L7R1DL6-ZjOIzrpAkLS:b-h95rsz>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/4d8c3b100b20feee24dc683f55dc4bcd79e63d4f/pkg/sql/logictest/testdata/logic_test/inverted_index#L520>):*
>
>
> query IT
> SELECT * from update_test where j @> '["b"]';
>
> This test was failing with the previous code.
> ------------------------------
>
> *pkg/sql/sqlbase/rowwriter.go, line 837 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r4-837:-L75jobQC_deeIwfBkZb:b-k4ez7h>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/0fd1ea4eebb73f35c52bc72b844fec6ed8f106bf/pkg/sql/sqlbase/rowwriter.go#L837>):*
>
> if len(oldSecondaryIndexEntries) <= i {
> batch.InitPut(newSecondaryIndexEntry.Key, &newSecondaryIndexEntry.Value,
> true)
> }
> I don't think so because there could be less new entries than old
>
> ------------------------------
>
> *pkg/sql/sqlbase/rowwriter.go, line 854 at r5
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L7R1R40VqLsnE0q4AAV:-L7R1R40VqLsnE0q4AAW:bw4euk7>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/4d8c3b100b20feee24dc683f55dc4bcd79e63d4f/pkg/sql/sqlbase/rowwriter.go#L854>):*
>
> 		// Do not update Indexes in the DELETE_ONLY state.
> 		if _, ok := ru.deleteOnlyIndex[i]; !ok {
>
> I'm not quite sure what to do about this, because if we have multiple
> inverted indexes all of their entries are going to be mixed at the end.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/23318>*
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/23318#issuecomment-372462953>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBIsQaHC6TlEGTszzwXdrfPMhSsFjks5tduOzgaJpZM4SaM-0>
> .
>

m-schneider commented:
I'm not sure that would work. Unless we specifically look at prefixes of the key in the KV, we still won't be able to tell if there are entries for multiple inverted indexes at the end of the list where they would have different delete only states. If I do separate them out as a list of lists, would we be able to cherry pick a refactoring of that size?

---







*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23318#-:-L7UowO-1N2HFogRXr6Z:b-bv5due)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
This needs to be cherry-picked into 2.0 when merged.

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -842,24 +842,35 @@ func (ru *RowUpdater) UpdateRow(
+
 		var expValue interface{}
-		if !bytes.Equal(newSecondaryIndexEntry.Key, secondaryIndexEntry.Key) {
+		if !bytes.Equal(newSecondaryIndexEntry.Key, oldSecondaryIndexEntry.Key) && len(oldSecondaryIndexEntry.Key) > 0 {
```
you should not need len(oldSecondaryIndexEntry.Key) > 0

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -842,24 +842,35 @@ func (ru *RowUpdater) UpdateRow(
+		oldSecondaryIndexEntry := oldSecondaryIndexEntries[i]
+		newSecondaryIndexEntry := newSecondaryIndexEntries[i]
+
+		inverted := index.Type == IndexDescriptor_INVERTED
```
I think we should have a code path here for inverted and at the end just ```continue```

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -714,7 +714,7 @@ func (ru *RowUpdater) UpdateRow(
 		for i := range newSecondaryIndexEntries {
-			if !bytes.Equal(newSecondaryIndexEntries[i].Key, secondaryIndexEntries[i].Key) {
+			if !bytes.Equal(newSecondaryIndexEntries[i].Key, oldSecondaryIndexEntries[i].Key) {
 				ru.Fks.addCheckForIndex(ru.Helper.Indexes[i].ID)
```
I'm pretty sure this is going to cause a panic on an out of index

vivekmenezes replied:
I also checked you do not need to call ru.Fks.addCheckForIndex() for inverted indexes

vivekmenezes replied:
That being said it would be nice to create a schema with a FK relationship between table A -> B say on a column like ID with an index on ID and with B having another json column on it. And running through some CASCADE ON DELETE/UPDATE to make sure that it doesn't mess up.

vivekmenezes replied:
how can you guarantee that the index i is not gonna cause a panic when used with ru.Helper.Indexes[i]

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -871,6 +882,16 @@ func (ru *RowUpdater) UpdateRow(
+		batch.Del(oldSecondaryIndexEntries[i].Key)
+	}
+
+	// We're adding all of the inverted index entries from the row being updated.
```
putFn = insertInvertedPutFn and use that below

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -871,6 +882,16 @@ func (ru *RowUpdater) UpdateRow(
 
+	// We're removing all of the inverted index entries from the row being updated.
+	for i := len(ru.Helper.Indexes); i < len(oldSecondaryIndexEntries); i++ {
+		batch.Del(oldSecondaryIndexEntries[i].Key)
```
if traceKV {
 			log.VEventf(ctx, 2, "Del %s", keys.PrettyPrint(ru.Helper.secIndexValDirs[i], oldSecondaryIndexEntry.Key))
  			}

although I don't know if ru.Helper.secIndexValDirs[i] is legal?

vivekmenezes replied:
I was suggesting you add the log message for debuggability

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -842,24 +842,35 @@ func (ru *RowUpdater) UpdateRow(
 		}
 		if traceKV {
 			log.VEventf(ctx, 2, "CPut %s -> %v", keys.PrettyPrint(ru.Helper.secIndexValDirs[i], newSecondaryIndexEntry.Key), newSecondaryIndexEntry.Value.PrettyPrint())
 		}
```
only applies to the CPUT codepath

vivekmenezes reviewed (commented):
Can you add some more unittests to beaf this up a bit?

m-schneider commented:
*[pkg/sql/sqlbase/rowwriter.go, line 718 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r6-718:-L7eJgD29NgxH0ZjuS0E:bx8kplw) ([raw file](https://github.com/cockroachdb/cockroach/blob/df72e2d23c329cbdd5d351388b739de7e9d5f951/pkg/sql/sqlbase/rowwriter.go#L718)):*
> addCheckForIndex
I added some tests for that and more tests for updates and upserts as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 0 of 2 files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/sql/sqlbase/rowwriter.go, line 852 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r6-852:-L7eColD2oeuzlsw4XWk:bkh6hd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/df72e2d23c329cbdd5d351388b739de7e9d5f951/pkg/sql/sqlbase/rowwriter.go#L852)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I think we should have a code path here for inverted and at the end just ```continue```
</blockquote></details>

It was only used in one place so I moved the condition down.

---

*[pkg/sql/sqlbase/rowwriter.go, line 855 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r6-855:-L7eCk7n84UMbkqerziW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df72e2d23c329cbdd5d351388b739de7e9d5f951/pkg/sql/sqlbase/rowwriter.go#L855)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

you should not need len(oldSecondaryIndexEntry.Key) > 0
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/rowwriter.go, line 868 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r6-868:-L7eCzMaK3nsOfuCPhbB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df72e2d23c329cbdd5d351388b739de7e9d5f951/pkg/sql/sqlbase/rowwriter.go#L868)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

only applies to the CPUT codepath
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/rowwriter.go, line 887 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r6-887:-L7eD2AXE4wYHhVv1rD1:bctrk8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/df72e2d23c329cbdd5d351388b739de7e9d5f951/pkg/sql/sqlbase/rowwriter.go#L887)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

if traceKV {
 			log.VEventf(ctx, 2, "Del %s", keys.PrettyPrint(ru.Helper.secIndexValDirs[i], oldSecondaryIndexEntry.Key))
  			}

although I don't know if ru.Helper.secIndexValDirs[i] is legal?
</blockquote></details>

I'm pretty sure it is, they should be lining up at this point in the code. I used it in debugging and it was helpful.

---

*[pkg/sql/sqlbase/rowwriter.go, line 890 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r6-890:-L7eD_S6GNCdhJnfdEEM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df72e2d23c329cbdd5d351388b739de7e9d5f951/pkg/sql/sqlbase/rowwriter.go#L890)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

putFn = insertInvertedPutFn and use that below
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 2 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 690 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L7eSZrSro_8Z0jcVvYv:-L7eSZrSro_8Z0jcVvYw:b-2kjk3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd367d9b1eaeeb6abf273cdfb50274990d31c6d7/pkg/sql/logictest/testdata/logic_test/inverted_index#L690)):*
> ```
> 2  {"a": "b", "c": "d"}
> 3  ["b", "c"]
> 5  ["a", "b"]
> ```

Yay! hooray for the greater number of tests!

---

*[pkg/sql/sqlbase/rowwriter.go, line 718 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r6-718:-L7eSnj-GC7ZNITPwvgI:b-rhafmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/df72e2d23c329cbdd5d351388b739de7e9d5f951/pkg/sql/sqlbase/rowwriter.go#L718)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

> addCheckForIndex
I added some tests for that and more tests for updates and upserts as well.
</blockquote></details>

so if len(newSecondaryIndexEntries) > len(ru.Helper.Indexes[i]) this will crash if the keys change

---

*[pkg/sql/sqlbase/rowwriter.go, line 887 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r6-887:-L7eTYYK7xcx53gj8dsc:bxc6o3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/df72e2d23c329cbdd5d351388b739de7e9d5f951/pkg/sql/sqlbase/rowwriter.go#L887)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I'm pretty sure it is, they should be lining up at this point in the code. I used it in debugging and it was helpful.
</blockquote></details>

I'm asking you to add the log message here for debugging but with the correct helper

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/logictest/testdata/logic_test/inverted_index:
```diff
@@ -504,3 +504,187 @@ query T
+SELECT * from update_cascade_test ORDER BY update_cascade;
+----
+2  {"a": "b", "c": "d"}
+3  ["b", "c"]
```
Yay! I'm glad for these tests! excellent!

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -842,24 +842,36 @@ func (ru *RowUpdater) UpdateRow(
+	for i, index := range ru.Helper.Indexes {
+		oldSecondaryIndexEntry := oldSecondaryIndexEntries[i]
+		newSecondaryIndexEntry := newSecondaryIndexEntries[i]
+
```
can you not change the code below and instead use

if index.Type == IndexDescriptor_INVERTED {
.....
continue
}

that way the inverted case is separated out and is easier to understand and you can add a comment for why you're deleting the old and appending the new value in this special code. Thanks!

m-schneider commented:
Review status: 0 of 3 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/inverted_index, line 551 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L7ew0c3kPQiYYMjZVAo:-L7ew0c3kPQiYYMjZVAp:b-6fayos) ([raw file](https://github.com/cockroachdb/cockroach/blob/42b37b6905b683ebba7a3bd7e4d10ab6dc077bf6/pkg/sql/logictest/testdata/logic_test/inverted_index#L551)):*
> ```
> 
> statement ok
> UPDATE update_test SET (i, j) = (10, '{"longKey1":"longValue1", "longKey2":"longValue2"}') WHERE i = 2;
> ```

This broke line 717 before.

---

*[pkg/sql/sqlbase/rowwriter.go, line 718 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r6-718:-L7egZsIX3HGaRe8IGl8:b-xw4nxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/df72e2d23c329cbdd5d351388b739de7e9d5f951/pkg/sql/sqlbase/rowwriter.go#L718)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

how can you guarantee that the index i is not gonna cause a panic when used with ru.Helper.Indexes[i]
</blockquote></details>

Good point, you also mentioned that we don't need to add those checks for inverted indexes, so I changed the bounds of the loop.

---

*[pkg/sql/sqlbase/rowwriter.go, line 887 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r6-887:-L7evezpI3FvaRRk69Km:b-iz8lny) ([raw file](https://github.com/cockroachdb/cockroach/blob/df72e2d23c329cbdd5d351388b739de7e9d5f951/pkg/sql/sqlbase/rowwriter.go#L887)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I was suggesting you add the log message for debuggability
</blockquote></details>

I'm not sure we can add an accurate log message here because if there are entries for multiple inverted indexes in this list we wouldn't be able to tell them apart.

---

*[pkg/sql/sqlbase/rowwriter.go, line 851 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r7-851:-L7egrv4y84aCKDO_Rl_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd367d9b1eaeeb6abf273cdfb50274990d31c6d7/pkg/sql/sqlbase/rowwriter.go#L851)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

can you not change the code below and instead use

if index.Type == IndexDescriptor_INVERTED {
.....
continue
}

that way the inverted case is separated out and is easier to understand and you can add a comment for why you're deleting the old and appending the new value in this special code. Thanks!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -842,33 +842,65 @@ func (ru *RowUpdater) UpdateRow(
+
+		// We're skipping inverted indexes in this loop, but appending the inverted index entry to the back of
+		// newSecondaryIndexEntries to process later. For inverted indexes we need to remove all old entries before adding
+		// new ones to prevent race conditions.
```
you can remove the "to prevent race conditions" part. It's not a race but the way operations need to be ordered.

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -842,33 +842,65 @@ func (ru *RowUpdater) UpdateRow(
+	if ru.DeleteHelper != nil {
+		for i, deletedSecondaryIndexEntry := range deleteOldSecondaryIndexEntries {
+			if i < len(ru.DeleteHelper.Indexes) {
+				ru.Fks.addCheckForIndex(ru.DeleteHelper.Indexes[i].ID, ru.DeleteHelper.Indexes[i].Type)
```
you can totally get rid of the need to check the index in the DELETE_ONLY state. So get rid of this line and thus the if statement above.

vivekmenezes reviewed (approved):
Great stuff! Please incorporate the suggestions. Thanks. Nice testing!

m-schneider commented:
Review status: 0 of 3 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/sql/sqlbase/rowwriter.go, line 854 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r9-854:-L7j7Km_lI5xj0YKa4vL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d35c3e54995215f287c3f6e33c733ab550b9a31c/pkg/sql/sqlbase/rowwriter.go#L854)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

you can remove the "to prevent race conditions" part. It's not a race but the way operations need to be ordered.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/rowwriter.go, line 886 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318#-L6bfKYtEp_j4NbCAh1L-r9-886:-L7j8EBmGbM53_hEI9x-:b-egns32) ([raw file](https://github.com/cockroachdb/cockroach/blob/d35c3e54995215f287c3f6e33c733ab550b9a31c/pkg/sql/sqlbase/rowwriter.go#L886)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

you can totally get rid of the need to check the index in the DELETE_ONLY state. So get rid of this line and thus the if statement above.
</blockquote></details>

I got rid of the foreign key check and moved the if statement down for logging. We could get array out of bounds errors without it with tracekv on.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23318)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -842,33 +842,69 @@ func (ru *RowUpdater) UpdateRow(
+					// Passing nil to the keys.PrettyPrint function is valid, because if there are entries beyond the number of
+					// indexes that means they are inverted index entries which are all encoded in the ascending direction.
+					log.VEventf(ctx, 2, "Del %s", keys.PrettyPrint(nil, deletedSecondaryIndexEntry.Key))
+
```
I think for all the Del() (this one and the one below we should start using 

```
log.VEventf(ctx, 2, "Del %s", deletedSecondaryIndexEntry.Key)
```

We don't need any of this if then else logic which is looking wrong.

vivekmenezes commented on pkg/sql/sqlbase/rowwriter.go:
```diff
@@ -842,33 +842,69 @@ func (ru *RowUpdater) UpdateRow(
+	}
+
+	// We're removing all of the inverted index entries from the row being updated.
+	for i := len(ru.Helper.Indexes); i < len(oldSecondaryIndexEntries); i++ {
```
```
if traceKV {
log.VEventf(ctx, 2, "Del %s", oldSecondaryIndexEntries[i].Key)
}
```

vivekmenezes commented:
please merge and cherry pick this change so that it goes out on monday's build. Thanks!
